<center>
Obrada prirodnog jezika

Lemmatization

Lucija Dumančić

# MODEL TRAINING

In [1]:
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [2]:
%cd /content/drive/MyDrive/OPJ/Projekt

/content/drive/MyDrive/OPJ/Projekt


##Load train and validation data

In [3]:
# Load train arrays from files
vectorized_tokens_train_val = np.load('vectorized_tokens_train.npy')
vectorized_lemmas_train_val = np.load('vectorized_lemmas_train.npy')

In [4]:
# Used for splitting train and validation vectors
split_index = int(len(vectorized_tokens_train_val) * 0.8)

In [5]:
# Train data
vectorized_tokens_train = vectorized_tokens_train_val[:split_index]
vectorized_lemmas_train = vectorized_lemmas_train_val[:split_index]

In [6]:
# Validation data
vectorized_tokens_val = vectorized_tokens_train_val[split_index:]
vectorized_lemmas_val = vectorized_lemmas_train_val[split_index:]

In [7]:
print(len(vectorized_tokens_train))
print(len(vectorized_tokens_val))

439554
109889


We utilized 439,554 vectorized tokens and lemmas for training the model, and the remaining portion, consisting of 109,889 vectorized tokens and lemmas, was allocated for validation purposes.

## Model

In [ ]:
# Define and compile the model
model = Sequential([
    Dense(128, activation='relu', input_shape=(vectorized_tokens_train.shape[1],)),   # Input layer with 128 neurons and ReLU activation function
    Dense(64, activation='relu'),   # Hidden layer with 64 neurons and ReLU activation function
    Dense(vectorized_lemmas_train.shape[1])   # Output layer with a number of neurons equal to the dimensionality of the output (lemmas)
])
model.compile(optimizer='adam', loss='cosine_similarity')   # Cosine similarity loss function is used for regression tasks

In this code cell, a neural network model is defined using the Keras Sequential API. It comprises three Dense layers: an input layer with 128 neurons utilizing ReLU activation, a hidden layer with 64 neurons also employing ReLU activation, and an output layer with a dynamic number of neurons equal to the dimensionality of the output (lemmas). The model is compiled with the Adam optimizer and utilizes cosine similarity as the loss function, making it suitable for regression tasks, particularly for predicting vectors closer in direction to the target vectors.

In [ ]:
# Train the model
model.fit(vectorized_tokens_train, vectorized_lemmas_train,
          validation_data=(vectorized_tokens_val, vectorized_lemmas_val), epochs=20, batch_size=64)

Epoch 1/20
6869/6869 [==============================] - 24s 3ms/step - loss: -0.7255 - val_loss: -0.7251
Epoch 2/20
6869/6869 [==============================] - 21s 3ms/step - loss: -0.7467 - val_loss: -0.7292
Epoch 3/20
6869/6869 [==============================] - 22s 3ms/step - loss: -0.7496 - val_loss: -0.7310
Epoch 4/20
6869/6869 [==============================] - 21s 3ms/step - loss: -0.7526 - val_loss: -0.7338
Epoch 5/20
6869/6869 [==============================] - 22s 3ms/step - loss: -0.7539 - val_loss: -0.7333
Epoch 6/20
6869/6869 [==============================] - 22s 3ms/step - loss: -0.7546 - val_loss: -0.7349
Epoch 7/20
6869/6869 [==============================] - 22s 3ms/step - loss: -0.7559 - val_loss: -0.7359
Epoch 8/20
6869/6869 [==============================] - 22s 3ms/step - loss: -0.7565 - val_loss: -0.7337
Epoch 9/20
6869/6869 [==============================] - 21s 3ms/step - loss: -0.7568 - val_loss: -0.7365
Epoch 10/20
6869/6869 [==============================] 

The model is trained using the fit method. Training data vectorized_tokens_train and corresponding target data vectorized_lemmas_train are provided for learning. Additionally, validation data (vectorized_tokens_val, vectorized_lemmas_val) is used to evaluate the model's performance during training. The training process lasts for 20 epochs with a batch size of 64 samples. During each epoch, the model adjusts its parameters to minimize the specified loss function, optimizing its ability to predict lemmas based on tokenized input.

In [ ]:
#Save and export lemmatization model
model.save("lemmatization_model.keras")